src/main/java/org/javarosa/xpath/analysis/XPathAccumulatingAnalyzer.java

```diff
@@ -0,0 +1,106 @@
+ * wherever it is present in the expression.
+ *
+ * IMPORTANT NOTE: An accumulating analyzer may analyze the same sub-expression or context ref of
+ * an expression multiple times in a single analysis pass. This means:
+ * - An AccumulatingAnalyzer is NOT appropriate to use for answering questions such as
+ * "How many times is X referenced in this expression?"
+ * - An accumulating analyzer IS appropriate to use for answering questions such as
+ * "What is the set of all things of X type which are referenced at least one time in this expression?"
+ *
+ * @author Aliza Stone
+ */
+public abstract class XPathAccumulatingAnalyzer<T> extends XPathAnalyzer {
+
+    protected List<T> accumulatedList;
+
+    public XPathAccumulatingAnalyzer() {
+        super();
+    }
+
+    public XPathAccumulatingAnalyzer(TreeReference contextRef) {
+        super(contextRef);
+    }
+
+    @Nullable
+    public Set<T> accumulate(XPathAnalyzable rootExpression) {
+        try {
+            rootExpression.applyAndPropagateAnalyzer(this);
+            Set<T> set = new HashSet<>();
+            for (T item : aggregateResults(new ArrayList<T>())) {
+                set.add(item);
+            }
+            return set;
+        } catch (AnalysisInvalidException e) {
+            return null;
+        }
+    }
+
+    private List<T> aggregateResults(List<T> aggregated) {
+        aggregated.addAll(this.accumulatedList);
+        for (XPathAnalyzer subAnalyzer : this.subAnalyzers) {
+            ((XPathAccumulatingAnalyzer)subAnalyzer).aggregateResults(aggregated);
+        }
+        return aggregated;
+    }
+
+    // For all AccumulatingAnalyzers, it should be sufficient to handle a current() reference by
+    // applying the analyzer separately to the expression itself and to the context ref
+    @Override
+    public void doAnalysisForTreeRefWithCurrent(TreeReference expressionWithContextTypeCurrent)
+            throws AnalysisInvalidException {
+
+        if (getOriginalContextRef() == null) {
+            throw new AnalysisInvalidException();
+        }
+
+        doNormalTreeRefAnalysis(expressionWithContextTypeCurrent);
+        // TODO: There may be a way to figure out more carefully under which circumstances
+        // getOriginalContextRef() actually represents something new to analyze (similar to the
+        // check below in doAnalysisForRelativeTreeRef), but for now this seems hard to determine
+        getOriginalContextRef().applyAndPropagateAnalyzer(this);
+    }
+
+    // For all AccumulatingAnalyzers, it should be sufficient to handle a relative reference by
+    // applying the analyzer separately to the expression itself and to the context ref
+    @Override
+    public void doAnalysisForRelativeTreeRef(TreeReference expressionWithContextTypeRelative)
+            throws AnalysisInvalidException {
+
+        if (getContextRef() == null) {
+            throw new AnalysisInvalidException();
```

Comment:
I think it would be useful to add messages to these, since it will make it easier to track down false negatives (IE: We get back an invalid analysis: Since the process is recursive it'll be really tricky to figure out where and what ended up throwing the issue.

Reply:
Good point, added.